### PR TITLE
fun/dirs: Fix $HOME selection in mcd and cd with a number 

### PR DESCRIPTION
**Pull request by pghvlaans**
In `fun/dirs`, the `mcd` function is broken:

* When selecting a number, `\cd` is called with that number, resulting in failed or (very occasionally) incorrect directory changes.
* When selecting a directory under `$HOME` by name, `\cd` attempts to change directory to a directory starting with escaped `~`. The results are similar.

In addition, `cd` no longer interacts with the push stack. In the original state of `fun/dirs` (added to the repository in early 2001, but previously appearing in Bolsky and Korn), using `cd 5` would change to the directory in the fifth position on the push stack, for example. 

`fun/dirs`: In the function `_cd`, clear all positional parameters when `mcd` is called and a directory under `$HOME` is selected by name, and whenever `cd` is called with a number as the first parameter. This allows `\cd` to use `$dirs` as intended in such cases.

### CODE REVIEW

**Comment by pghvlaans** [comment]
~~As a side note, these commits make it so that e.g. `cd 5` or `cd -2` attempt to change directory to literal `5` and `-2`, respectively, rather than whatever is in position on the push stack.~~

~~I couldn't find any documentation from the time `dirs` was introduced (the first half of 2001) to indicate whether those kinds of `cd` calls should work, but enabling that behavior would be a simple change. Please let me know which would be preferable.~~

Never mind that; I tried reverting the most recent commit (6109f54), and `cd` with numbers interacted with the push stack.

**Comment by pghvlaans** [comment]
After a little more testing, it turned out that in the original `fun/dirs`, something like `cd 2 whatever` would always execute `cd 2`.

File: src/cmd/ksh93/fun/dirs [hunk 1]
```diff
@@ -30,16 +32,28 @@ function _cd
     case $1 in
     -|-1|2) # \cd -
         n=_push_top type=1
+        while (($#))
+        do
+              shift
+        done
```
Comment (McDutchie) [inline src/cmd/ksh93/fun/dirs]: The positional parameters can be emptied all at once using `set --`, which is more efficient than removing them one by one.

File: src/cmd/ksh93/fun/dirs [hunk 2]
```diff
@@ -55,6 +69,13 @@ function _cd
     case $dir in
     \~*)   dir=$HOME${dir#\~}
     esac
+    # mcd passes escaped ~, which must be dealt with
+    if     ((_mcd_called>0))
+    then   case "$@" in
```
Comment (McDutchie) [inline src/cmd/ksh93/fun/dirs]: Using `"$@"` in this context makes no sense as that quoted expansion generates one word per positional parameter, but only exactly one word is possible between `case` and `in`, so the behaviour of `"$@"` is unspecified in this context. On ksh, in practice it's the same as `"$*"` here, i.e., the positional parameters get joined into one string separated by the first character of `$IFS`. But in this case you're only checking the first character of the first parameter, so you might as well just use `$1` (no quotes needed).

File: src/cmd/ksh93/fun/dirs (hunk 1 above)
Comment (pghvlaans) [reply src/cmd/ksh93/fun/dirs]: That makes sense. Thank you.

File: src/cmd/ksh93/fun/dirs (hunk 2 above)
Comment (pghvlaans) [reply src/cmd/ksh93/fun/dirs]: Yes, only the first parameter would be wanted here.

**Pull request merged**